### ISSUE TITLE
ui (histograms): Render Native Histograms in Graph view

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#11268]
The text rendering of Native Histograms in the Table view is quite verbose, but it's currently the only way of displaying Native Histograms in the built-in UI.

It would be cool to also display them in the Graph view rather than only in the Table view. The most obvious way of representing them would be a heat map, but maybe there are others.

A heat map can only show one histogram series at a time, so a query that yields multiple histogram series needed to produce a number of heat maps, which is distinctly different from what we do so far, namely redering many graphs in a single panel. Plus, there is the case of mixed histogram and conventional float series, for which we probably would need to render the conventional graph panel followed by _n_ heat maps.

**Comment by beorn7** [comment]
Maybe we can render multiple heat maps in one panel after all. Options:
- Rather than rendering one heat map with a "rainbow" style color schema, we could render it in one color from dark to light. Then we could overlay multiple heatmaps, each in a different color. With the right transparency, this might be actually useful (cf. astronomical pictures where very different parts of the spectrum, e.g. infrared, xray, are mapped to different colors). This would, in a way, correspond to how we render different time series with a different color each.
- Rotate through the different heat maps over time.

I can imagine that each mode of rendering (side-by-side, color-coded, rotate-over-time) has its pros and cons and it depends on the concrete case which one will fare best. So perhaps give the user an option to easily switch.

**Comment by bboreham** [comment]
Is this covered by #13096 ?

**Comment by beorn7** [comment]
#13096 is only for classic histograms.

**Comment by beorn7** [comment]
#13680 is still being worked on to get the table view right for native histograms. Once that is done, it should be relatively straight-forward to extend the approach to the graph view AKA heat maps.

**Comment by anshjain16** [comment]
Hey @beorn7 , Is anyone working on this issue

**Comment by kushalShukla-web** [comment]
hey @beorn7  im also interested , can you please assign this to me ?

**Comment by Maniktherana** [comment]
Hey @anshjain16 feel free to check out #13680. If you can help out there Id be happy to collaborate with you on this issue

**Comment by anshjain16** [comment]
> Hey @anshjain16 feel free to check out #13680. If you can help out there Id be happy to collaborate with you on this issue



> Hey @anshjain16 feel free to check out #13680. If you can help out there Id be happy to collaborate with you on this issue

Yeah sure, i would love to contribute, let me just get up to date with what all things have been covered in the issue and then I'll be good to go, till then what means do you recommend to communicate to you if I have any doubts regarding the same?

**Comment by beorn7** [comment]
#13680 is finally done. @Maniktherana plans to tackle heatmaps (i.e. this issue) next. @anshjain16 if you want to help out, you are welcome. Please get in touch with @Maniktherana .

**Comment by alexgreenbank** [comment]
Hello from the bug scrub.

This is still in progress.

**Comment by Maniktherana** [comment]
Yeah Ive picked it up but got a little side tracked

**Comment by beorn7** [comment]
I just noticed that range vectors are not rendered at all now:

![image](https://github.com/user-attachments/assets/c91afc44-4914-4b35-bbff-f64397ac0abd)

In the makeshift rendering we had before the graphical rendering, we at least had the output of `FloatHistogram.String()` rendered, together with the usual `@`-timestamp notation. It was ugly, but still helpful when per-sample inspections of the TSDB is required.

I think we have to bring something similar back. Graphical rendering is probably too much for range vectors, but at the very least we can just do the ugly text rendering again. (Maybe we could also do a heatmap, like we plan for the Graph view, but the textview should still be an option for low-level inspection.)

**Comment by beorn7** [comment]
WRT heatmaps, I wanted to write down some finer points that I've had in mind for a while:

- A heatmap is like many histograms lined up from left to right over time, each rotated 90 degrees counter-clockwise, and bucket population now rendered as colors rather than bars of a specific height. Let's call each "bar" a "tile" in the heatmap.
- What used to be the _x_ axis in a histogram rendering, is now the _y_ axis in the heatmap, with exactly the same questions about exponential or linear scale of the axis and what that means for the height of each tile (formerly called "width of the bar").
- Like the "height of the bar" before, we have to normalize the color based on the population in the bucket divided by the height of the tile. For example, a tile with 100 bucket population must have the same color as a tile with 200 bucket population but double height (i.e. the bucket has twice the width, if we have linear rendering).
- An interesting challenge is that a histogram might change resolution over time, and buckets might appear and disappear over time. All of this has to arrange properly in the rendering, i.e. the heatmap does not necessarily come out as a regular grid of tiles.
- As with the normal histogram already implemented, the zero bucket and negative bucket have to be treated properly.

Grafana already renders beautiful heatmaps, based on various data sources, including Prometheus. The demos you might have seen during talks so far were done with Grafana. The implementation might serve as an inspiration, but I am not sure it already handles all the aspects above properly, in particular when it comes to resolution changes (and as a related topic non-standard schemas like the upcoming NHCB) and to correct color normalization if the tile height changes. I have also never tried what happens in Grafana if there are negative buckets or a populated zero bucket (including the case where it has a width of zero).

An interesting point is how to render the case where the query results in multiple histogram series. The naive approach (and certainly the one we should go for first) is to render many heatmaps side-by-side, maybe in a grid "as square as possible". This might be very resource demanding, but let's see how for we can go with it.

But there might be more interesting way of doing it:
- Have only one heatmap panel, but cycle through the different histogram series over time, i.e. switch from one to the next every second or so, with buttons to stop the cycler and select and iterate through the various series via push of buttons. It will create some kind of mini movie. Humans are pretty good in spotting differences between pictures if presented in this way.
- Give every histogram series its own color (e.g. red, green, blue), render a heatmap for each by "color coding" the bucket population just along the brightness scale for the respective color. Then put them all into the same panel as overlays. This will only work with a limited number of histogram series, but could be very powerful, leveraging human sense of color. Compare this to astronomical pictures where different parts of the spectrum (IR, UV, gamma, x-ray) are mapped to different colors (i.e. mapped to the visible spectrum) and overlaid to a color picture, e.g. [here](https://en.wikipedia.org/wiki/Messier_94#/media/File:Small_new_m94_announcement.jpg).

And here is yet another idea of coloring a (single) heatmap (also for later consideration, I just wanted to write it down):

What Grafana does, and what's described so far above is coloring according to bucket population (normalized by height of the tile). Now there is a different way, that could also be very instructive:
- Imagine we have the conventional Prometheus _Graph_ view.
- As we cannot render a heatmap (yet), e.g. directly using a query like `rate(rpc_latency_seconds[5m])`, we render the 75th percentile using a query like `histogram_quantile(rate(0.75, rpc_latency_seconds[5m]))`, resulting in a normal float series.
- We could render this query with a blue line.
- Then we can add another query to the graph, rendering it in green: `histogram_quantile(rate(0.5, rpc_latency_seconds[5m]))`.
- Then add another query, rendering it in red: `histogram_quantile(0,25, rate(rpc_latency_seconds[5m]))`.
- And now you might get what I'm up to: Let's render (as a thought experiment) a hundred of these graphs, for each percentile from 0 to 100 (or 0.00 to 1.00 as the `histogram_quantile` parameter) and give it a color from the rainbow.
- What we get is a heatmap, where the same color corresponds to the same percentile.

I think such a rendering would be good if latency percentiles are in the center of interest, but not so much how much traffic is currently served. With the conventional coloring, low traffic times would look very "dark" or "cold" (or whatever color mapping you use), while the heatmap lights up with more traffic. With my proposed alternative coloring, your heatmap would always be "hot" at the bottom and "cold" at the top, and depending on your latency distribution, the "heat would rise" depending on how bad it is. E.g. if only outliers have high latency, most of the area will still be pretty "cool", while a high latency for many requests will fill most of the area with a "hot" color.

**Comment by Maniktherana** [comment]
I should get back on this. Will give it a shot over the weekend

**Comment by amanycodes** [comment]
Hi @Maniktherana are you still working on this issue?

**Comment by beorn7** [comment]
No news from @Maniktherana in a while. Unassigning…
